File: server/game/AIPlayer.js
const { Commands } = require('../../shared/commands');
const CONSTANTS = require('../../shared/constants');

// AI Player that makes decisions and issues commands
class AIPlayer {
  constructor(playerId, gameLoop, playerIndex) {
    this.playerId = playerId;
    this.gameLoop = gameLoop;
    this.playerIndex = playerIndex;
    this.tickCount = 0;
    this.decisionInterval = 30; // Make decisions every 30 ticks (~0.5 sec)

    // Track what workers are assigned to
    this.workerAssignments = new Map(); // workerId -> 'gathering' | 'building' | 'idle'

    // Track pending builds
    this.pendingBarracks = false;
    this.pendingSupplyDepot = false;
  }

  // Called every game tick
  update() {
    this.tickCount++;

    // Only make decisions periodically to avoid spam
    if (this.tickCount % this.decisionInterval !== 0) {
      return;
    }

    const world = this.gameLoop.world;
    const playerState = this.gameLoop.players.get(this.playerId);
    if (!playerState) return;

    // Get my units and buildings
    const myUnits = this.getMyUnits(world);
    const myBuildings = this.getMyBuildings(world);
    const myWorkers = myUnits.filter(u => u.subtype === 'worker');
    const mySoldiers = myUnits.filter(u => u.subtype === 'soldier');
    const myBase = myBuildings.find(b => b.subtype === 'base');
    const myBarracks = myBuildings.filter(b => b.subtype === 'barracks' && b.state !== 'constructing');
    const mySupplyDepots = myBuildings.filter(b => b.subtype === 'supplyDepot');

    // Get resources and ball
    const resources = this.getResources(world);
    const ball = world.getBall();

    // Decision priority:
    // 1. Assign idle workers to gather
    // 2. Build supply depot if supply blocked
    // 3. Build barracks if we don't have one
    // 4. Train workers (up to 8)
    // 5. Train soldiers
    // 6. Send soldiers to push ball toward enemy goal

    // 1. Assign idle workers to gather
    this.assignWorkersToGather(myWorkers, resources, world);

    // 2. Build supply depot if supply blocked (or close to it)
    const supplyBuffer = 2;
    const needSupply = playerState.supply + supplyBuffer >= playerState.maxSupply;
    const constructingSupply = myBuildings.some(b => b.subtype === 'supplyDepot' && b.state === 'constructing');

    if (needSupply && !constructingSupply && playerState.minerals >= 100) {
      this.buildStructure(myWorkers, myBase, 'supplyDepot', world);
    }

    // 3. Build barracks if we don't have one
    const hasBarracks = myBarracks.length > 0;
    const constructingBarracks = myBuildings.some(b => b.subtype === 'barracks' && b.state === 'constructing');

    if (!hasBarracks && !constructingBarracks && playerState.minerals >= 150) {
      this.buildStructure(myWorkers, myBase, 'barracks', world);
    }

    // 4. Train workers (up to 8)
    if (myWorkers.length < 8 && myBase && myBase.state !== 'constructing') {
      const workerCost = 50;
      const workerSupply = 1;
      if (playerState.minerals >= workerCost && playerState.supply + workerSupply <= playerState.maxSupply) {
        // Check if not already training
        const trainingQueue = myBase.trainingQueue || [];
        if (trainingQueue.length < 2) {
          this.queueCommand(Commands.train(myBase.id, 'worker'));
        }
      }
    }

    // 5. Train soldiers from barracks
    for (const barracks of myBarracks) {
      const soldierCost = 100;
      const soldierSupply = 2;
      if (playerState.minerals >= soldierCost && playerState.supply + soldierSupply <= playerState.maxSupply) {
        const trainingQueue = barracks.trainingQueue || [];
        if (trainingQueue.length < 2) {
          this.queueCommand(Commands.train(barracks.id, 'soldier'));
        }
      }
    }

    // 6. Send soldiers to push ball toward enemy goal
    if (mySoldiers.length > 0 && ball) {
      this.commandSoldiersToPushBall(mySoldiers, ball, world);
    }
  }

  getMyUnits(world) {
    return world.getAllActors().filter(a =>
      a.type === 'unit' && a.ownerId === this.playerId
    );
  }

  getMyBuildings(world) {
    return world.getAllActors().filter(a =>
      a.type === 'building' && a.ownerId === this.playerId
    );
  }

  getResources(world) {
    return world.getAllActors().filter(a => a.type === 'resource');
  }

  assignWorkersToGather(workers, resources, world) {
    if (resources.length === 0) return;

    for (const worker of workers) {
      // Skip workers already gathering or building
      if (worker.state === 'gathering' || worker.state === 'returning' || worker.state === 'building') {
        continue;
      }

      // Find nearest resource
      let nearestResource = null;
      let nearestDist = Infinity;

      for (const resource of resources) {
        if (resource.amount <= 0) continue;
        const dx = resource.x - worker.x;
        const dy = resource.y - worker.y;
        const dist = Math.sqrt(dx * dx + dy * dy);
        if (dist < nearestDist) {
          nearestDist = dist;
          nearestResource = resource;
        }
      }

      if (nearestResource) {
        this.queueCommand(Commands.gather([worker.id], nearestResource.id));
      }
    }
  }

  buildStructure(workers, base, buildingType, world) {
    if (!base) return;

    // Find an idle or gathering worker
    const availableWorker = workers.find(w =>
      w.state === 'idle' || w.state === 'gathering' || w.state === 'returning'
    );

    if (!availableWorker) return;

    // Calculate build position (away from base toward center)
    const worldCenterX = (world.width * CONSTANTS.TILE_WIDTH) / 2;
    const directionX = worldCenterX > base.x ? 1 : -1;

    // Build position offset from base
    const offsetX = directionX * 120;
    const offsetY = buildingType === 'supplyDepot' ? -80 : 80; // Supply above, barracks below

    const buildX = base.x + offsetX;
    const buildY = base.y + offsetY;

    this.queueCommand(Commands.build(availableWorker.id, buildingType, buildX, buildY));
  }

  commandSoldiersToPushBall(soldiers, ball, world) {
    // Determine enemy goal position (opposite side of map)
    const worldWidth = world.width * CONSTANTS.TILE_WIDTH;
    const worldHeight = world.height * CONSTANTS.TILE_HEIGHT;
    const goalY = worldHeight / 2;

    // AI is playerIndex 1, so enemy goal is on the left (x = 0)
    // AI is playerIndex 0, so enemy goal is on the right (x = worldWidth)
    const enemyGoalX = this.playerIndex === 0 ? worldWidth : 0;

    // Get soldiers that aren't already attacking the ball
    const availableSoldiers = soldiers.filter(s =>
      s.state === 'idle' || s.state === 'moving' ||
      (s.state === 'attacking' && s.attackTargetId !== ball.id)
    );

    for (const soldier of availableSoldiers) {
      const dx = ball.x - soldier.x;
      const dy = ball.y - soldier.y;
      const distToBall = Math.sqrt(dx * dx + dy * dy);

      // Calculate position behind ball (relative to enemy goal) to push it in the right direction
      const ballToGoalX = enemyGoalX - ball.x;
      const ballToGoalY = goalY - ball.y;
      const ballToGoalDist = Math.sqrt(ballToGoalX * ballToGoalX + ballToGoalY * ballToGoalY);

      // If soldier is on the correct side of the ball (between ball and our goal), attack the ball
      // This makes them push it toward the enemy goal
      const soldierToBallX = ball.x - soldier.x;
      const dotProduct = soldierToBallX * ballToGoalX + (ball.y - soldier.y) * ballToGoalY;
      const isOnCorrectSide = dotProduct > 0;

      if (isOnCorrectSide && distToBall < 400) {
        // Attack the ball - this will make the soldier move toward it and continuously push it
        this.queueCommand(Commands.attack([soldier.id], ball.id));
      } else if (ballToGoalDist > 0) {
        // Move to position behind the ball (relative to enemy goal)
        const pushDistBehind = 200;
        const targetX = ball.x - (ballToGoalX / ballToGoalDist) * pushDistBehind;
        const targetY = ball.y - (ballToGoalY / ballToGoalDist) * pushDistBehind;

        this.queueCommand(Commands.move([soldier.id], targetX, targetY));
      }
    }
  }

  queueCommand(command) {
    this.gameLoop.queueCommand(this.playerId, command);
  }
}

module.exports = AIPlayer;

File: server/AI/AIPlayer.cs
using server.ECS;
using server.ECS.Components.Core;
using server.ECS.Components.Combat;
using server.ECS.Components.Movement;
using server.ECS.Components.Unit;
using server.ECS.Components.Building;
using server.ECS.Components.Avatar;
using server.ECS.Components.Resource;
using server.ECS.Messages;
using server.Rooms;
using server.Setup;

namespace server.AI;

/// <summary>
/// AI player that makes decisions and sends commands like a human player.
/// Runs as part of the game loop.
/// </summary>
public class AIPlayer
{
    private readonly string _playerId;
    private readonly int _playerIndex;
    private readonly World _world;
    private readonly PlayerState _playerState;
    private readonly string _aiType;

    // AI tuning parameters
    private const int MAX_WORKERS = 8;
    private const int SUPPLY_BUFFER = 3;  // Build depot when supply within this of max
    private const float DECISION_INTERVAL = 0.5f;  // Seconds between major decisions

    private float _decisionTimer = 0f;
    private int _enemyPlayerIndex;

    public AIPlayer(string playerId, int playerIndex, World world, PlayerState playerState, string aiType = "normal")
    {
        _playerId = playerId;
        _playerIndex = playerIndex;
        _world = world;
        _playerState = playerState;
        _aiType = aiType;
        _enemyPlayerIndex = playerIndex == 0 ? 1 : 0;
    }

    /// <summary>
    /// Called every tick to let AI make decisions.
    /// </summary>
    public void Tick(float deltaTime)
    {
        // Empty AI does nothing
        if (_aiType == "empty")
            return;

        _decisionTimer += deltaTime;

        // Always control avatar (push ball toward enemy goal)
        ControlAvatar();

        // Make strategic decisions periodically
        if (_decisionTimer >= DECISION_INTERVAL)
        {
            _decisionTimer = 0f;
            MakeStrategicDecisions();
        }
    }

    private void ControlAvatar()
    {
        var avatar = FindAvatar();
        if (avatar == 0)
            return;

        var avatarTransform = _world.Get<Transform>(avatar);
        if (avatarTransform == null)
            return;

        var ball = FindBall();
        if (ball == 0)
            return;

        var ballTransform = _world.Get<Transform>(ball);
        if (ballTransform == null)
            return;

        // Calculate direction to push ball toward enemy goal
        var enemyGoalX = _playerIndex == 0 ? GameConstants.WORLD_PIXEL_WIDTH : 0f;
        var goalDirection = _playerIndex == 0 ? 1 : -1;

        // If ball is behind us, go around it
        var isBallBehindUs = (_playerIndex == 0 && ballTransform.x < avatarTransform.x - 50) ||
                            (_playerIndex == 1 && ballTransform.x > avatarTransform.x + 50);

        int dirX = 0;
        int dirY = 0;

        if (isBallBehindUs)
        {
            // Move toward ball first
            dirX = ballTransform.x > avatarTransform.x ? 1 : -1;
            dirY = ballTransform.y > avatarTransform.y ? 1 : (ballTransform.y < avatarTransform.y ? -1 : 0);
        }
        else
        {
            // Position ourselves to push ball toward enemy goal
            var dx = ballTransform.x - avatarTransform.x;
            var dy = ballTransform.y - avatarTransform.y;
            var dist = MathF.Sqrt(dx * dx + dy * dy);

            if (dist > 80)
            {
                // Move toward ball
                dirX = dx > 0 ? 1 : -1;
                dirY = dy > 20 ? 1 : (dy < -20 ? -1 : 0);
            }
            else
            {
                // We're near ball, push it toward goal
                dirX = goalDirection;
                dirY = 0;

                // Adjust Y to keep ball centered
                var centerY = GameConstants.WORLD_PIXEL_HEIGHT / 2f;
                if (ballTransform.y < centerY - 100)
                    dirY = 1;
                else if (ballTransform.y > centerY + 100)
                    dirY = -1;
            }
        }

        _world.Send(new AvatarMoveCommand
        {
            playerId = _playerId,
            directionX = dirX,
            directionY = dirY
        });
    }

    private void MakeStrategicDecisions()
    {
        // Priority order:
        // 1. Build supply depot if needed
        // 2. Train workers if we need more
        // 3. Build barracks if we don't have one
        // 4. Train soldiers
        // 5. Assign idle workers to gather

        var workers = GetUnits("worker");
        var soldiers = GetUnits("soldier");
        var workerCount = workers.Count;
        var soldierCount = soldiers.Count;

        var baseBuilding = FindBase();
        var barracks = FindBuilding("barracks");
        var supplyDepot = FindBuilding("supplyDepot");

        // Check if we're near supply cap
        var nearSupplyCap = _playerState.Supply >= _playerState.MaxSupply - SUPPLY_BUFFER;

        // 1. Build supply depot if needed and we have a worker
        if (nearSupplyCap && workerCount > 0)
        {
            var buildingSupplyDepot = FindBuildingUnderConstruction("supplyDepot");
            if (buildingSupplyDepot == 0)
            {
                var depotDef = EntityDefinitions.GetBuilding("supplyDepot")!;
                if (_playerState.CanAfford(depotDef.cost))
                {
                    var idleWorker = FindIdleWorker(workers);
                    if (idleWorker != 0)
                    {
                        var buildPos = GetBuildPosition();
                        _world.Send(new BuildCommand
                        {
                            playerId = _playerId,
                            workerId = idleWorker,
                            buildingType = "supplyDepot",
                            x = buildPos.x,
                            y = buildPos.y
                        });
                        return;
                    }
                }
            }
        }

        // 2. Train workers if we need more
        if (workerCount < MAX_WORKERS && baseBuilding != 0)
        {
            var workerDef = EntityDefinitions.GetUnit("worker")!;
            if (_playerState.CanAfford(workerDef.cost) && _playerState.CanSupport(workerDef.supply))
            {
                var queue = _world.Get<TrainingQueue>(baseBuilding);
                if (queue != null && queue.queue.Count == 0)
                {
                    _world.Send(new TrainRequestMessage
                    {
                        buildingEntity = baseBuilding,
                        unitType = "worker",
                        cost = workerDef.cost,
                        supply = workerDef.supply,
                        trainTime = workerDef.trainTime,
                        playerId = _playerId
                    });
                    return;
                }
            }
        }

        // 3. Build barracks if we don't have one
        if (barracks == 0 && workerCount > 0)
        {
            var buildingBarracks = FindBuildingUnderConstruction("barracks");
            if (buildingBarracks == 0)
            {
                var barracksDef = EntityDefinitions.GetBuilding("barracks")!;
                if (_playerState.CanAfford(barracksDef.cost))
                {
                    var idleWorker = FindIdleWorker(workers);
                    if (idleWorker != 0)
                    {
                        var buildPos = GetBuildPosition();
                        _world.Send(new BuildCommand
                        {
                            playerId = _playerId,
                            workerId = idleWorker,
                            buildingType = "barracks",
                            x = buildPos.x,
                            y = buildPos.y
                        });
                        return;
                    }
                }
            }
        }

        // 4. Train soldiers
        if (barracks != 0)
        {
            var soldierDef = EntityDefinitions.GetUnit("soldier")!;
            if (_playerState.CanAfford(soldierDef.cost) && _playerState.CanSupport(soldierDef.supply))
            {
                var queue = _world.Get<TrainingQueue>(barracks);
                if (queue != null && queue.queue.Count < 2) // Allow small queue
                {
                    _world.Send(new TrainRequestMessage
                    {
                        buildingEntity = barracks,
                        unitType = "soldier",
                        cost = soldierDef.cost,
                        supply = soldierDef.supply,
                        trainTime = soldierDef.trainTime,
                        playerId = _playerId
                    });
                }
            }
        }

        // 5. Assign idle workers to gather
        AssignIdleWorkersToGather(workers);

        // 6. Send soldiers toward ball
        SendSoldiersTowardBall(soldiers);
    }

    private void AssignIdleWorkersToGather(List<long> workers)
    {
        var minerals = FindMineralPatches();
        if (minerals.Count == 0)
            return;

        foreach (var worker in workers)
        {
            var state = _world.Get<UnitState>(worker);
            if (state == null || state.state != "idle")
                continue;

            // Skip workers that are being carried
            if (_world.Has<Carried>(worker))
                continue;

            // Skip auto-attack only workers (dropped by avatar)
            if (_world.Has<AutoAttackOnly>(worker))
                continue;

            // Find nearest mineral patch
            var workerTransform = _world.Get<Transform>(worker);
            if (workerTransform == null)
                continue;

            var nearestMineral = FindNearestMineral(workerTransform.x, workerTransform.y, minerals);
            if (nearestMineral != 0)
            {
                _world.Send(new GatherCommand
                {
                    playerId = _playerId,
                    workerIds = new List<long> { worker },
                    resourceId = nearestMineral
                });
            }
        }
    }

    private void SendSoldiersTowardBall(List<long> soldiers)
    {
        var ball = FindBall();
        if (ball == 0)
            return;

        var ballTransform = _world.Get<Transform>(ball);
        if (ballTransform == null)
            return;

        foreach (var soldier in soldiers)
        {
            var state = _world.Get<UnitState>(soldier);
            if (state == null)
                continue;

            // Only move idle soldiers
            if (state.state != "idle")
                continue;

            // Skip soldiers that are being carried or auto-attack only
            if (_world.Has<Carried>(soldier) || _world.Has<AutoAttackOnly>(soldier))
                continue;

            // Move toward ball
            _world.Send(new MoveCommand
            {
                playerId = _playerId,
                actorIds = new List<long> { soldier },
                targetX = ballTransform.x,
                targetY = ballTransform.y
            });
        }
    }

    #region Helper Methods

    private long FindAvatar()
    {
        foreach (var (entity, entityType) in _world.GetAll<EntityType>())
        {
            if (entityType.type != "avatar")
                continue;

            var ownership = _world.Get<Ownership>(entity);
            if (ownership?.ownerId == _playerId)
                return entity;
        }
        return 0;
    }

    private long FindBall()
    {
        foreach (var (entity, entityType) in _world.GetAll<EntityType>())
        {
            if (entityType.type == "ball")
                return entity;
        }
        return 0;
    }

    private long FindBase()
    {
        return FindBuilding("base");
    }

    private long FindBuilding(string subtype)
    {
        foreach (var (entity, entityType) in _world.GetAll<EntityType>())
        {
            if (entityType.type != "building" || entityType.subtype != subtype)
                continue;

            var ownership = _world.Get<Ownership>(entity);
            if (ownership?.ownerId != _playerId)
                continue;

            // Check if complete (no Construction component)
            if (_world.Has<Construction>(entity))
                continue;

            return entity;
        }
        return 0;
    }

    private long FindBuildingUnderConstruction(string subtype)
    {
        foreach (var (entity, entityType) in _world.GetAll<EntityType>())
        {
            if (entityType.type != "building" || entityType.subtype != subtype)
                continue;

            var ownership = _world.Get<Ownership>(entity);
            if (ownership?.ownerId != _playerId)
                continue;

            if (_world.Has<Construction>(entity))
                return entity;
        }
        return 0;
    }

    private List<long> GetUnits(string subtype)
    {
        var units = new List<long>();
        foreach (var (entity, entityType) in _world.GetAll<EntityType>())
        {
            if (entityType.type != "unit" || entityType.subtype != subtype)
                continue;

            var ownership = _world.Get<Ownership>(entity);
            if (ownership?.ownerId == _playerId)
                units.Add(entity);
        }
        return units;
    }

    private long FindIdleWorker(List<long> workers)
    {
        foreach (var worker in workers)
        {
            var state = _world.Get<UnitState>(worker);
            if (state?.state == "idle" && !_world.Has<Carried>(worker) && !_world.Has<AutoAttackOnly>(worker))
                return worker;
        }
        return 0;
    }

    private List<long> FindMineralPatches()
    {
        var minerals = new List<long>();
        foreach (var (entity, entityType) in _world.GetAll<EntityType>())
        {
            if (entityType.type == "resource" && entityType.subtype == "minerals")
            {
                var amount = _world.Get<ResourceAmount>(entity);
                if (amount != null && amount.amount > 0)
                    minerals.Add(entity);
            }
        }
        return minerals;
    }

    private long FindNearestMineral(float x, float y, List<long> minerals)
    {
        long nearest = 0;
        var nearestDist = float.MaxValue;

        foreach (var mineral in minerals)
        {
            var transform = _world.Get<Transform>(mineral);
            if (transform == null)
                continue;

            var dx = transform.x - x;
            var dy = transform.y - y;
            var dist = MathF.Sqrt(dx * dx + dy * dy);

            if (dist < nearestDist)
            {
                nearestDist = dist;
                nearest = mineral;
            }
        }

        return nearest;
    }

    private (float x, float y) GetBuildPosition()
    {
        var baseEntity = FindBase();
        if (baseEntity == 0)
            return (GameConstants.WORLD_PIXEL_WIDTH / 2f, GameConstants.WORLD_PIXEL_HEIGHT / 2f);

        var baseTransform = _world.Get<Transform>(baseEntity);
        if (baseTransform == null)
            return (GameConstants.WORLD_PIXEL_WIDTH / 2f, GameConstants.WORLD_PIXEL_HEIGHT / 2f);

        // Build in a position near the base but toward center
        var offset = _playerIndex == 0 ? 150f : -150f;
        var buildX = baseTransform.x + offset;

        // Alternate Y position based on existing buildings
        var yOffset = (Random.Shared.Next(2) == 0 ? -1 : 1) * (100 + Random.Shared.Next(100));
        var buildY = baseTransform.y + yOffset;

        // Clamp to map bounds
        buildY = Math.Clamp(buildY, 100, GameConstants.WORLD_PIXEL_HEIGHT - 100);

        return (buildX, buildY);
    }

    #endregion
}
